Implement Dominated Novelty Search

ryanboldi opened the pull request:
## Description

Add Dominated Novelty Search (https://arxiv.org/abs/2502.00593) as an archive to be used in pyres

## TODO

- [x] Implement DNS
- [x] Minimal tests
- [x] Rigorous tests, potentially integration with non-gaussian emitters.
- [ ] Docs update
- [ ] Tutorial notebook

## Status

- [x] I have read the guidelines in
      [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md)
- [x] I have linted and formatted my code with `ruff` and `ty`
- [x] I have tested my code by running `pytest`
- [x] I have added a description of my change to the changelog in `HISTORY.md`
- [x] This PR is ready to go

btjanaka commented on tests/archives/dns_archive_test.py:
```diff
@@ -0,0 +1,67 @@
+
+    # now add 10 more solutions, cap should stay at 10
+    archive.add(
+        np.random.Generator(np.random.PCG64()).random((10, 2)),
```
At the start of this test function, create a random generator with `rng = np.random.default_rng(42)`, where 42 is your seed. Then, call `rng.random()` in this code.

btjanaka commented on tests/archives/dns_archive_test.py:
```diff
@@ -0,0 +1,67 @@
+    assert archive.capacity == 10, "Capacity should stay at 10 when adding when full"
+
+
+# ensure that DNS removes duplicates
```
Can you make this a docstring for the function?

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,587 @@
+
+    ## Methods for writing to the archive ##
+
+    def _maybe_resize(self, new_size: int) -> None:
```
Just remove this method since it is not used?

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,587 @@
+        index of a potential new best elite (new_best_index).
+        """
+        _, new_best_elite_raw = self._store.retrieve([new_best_index])
+        new_best_elite_dict = cast(BatchData, new_best_elite_raw)
```
Any reason for casting here? I think you can just revert it to match what is done in the other archives.

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,587 @@
+        if measures.shape[0] == 0:
+            dns_scores = np.array([], dtype=self.dtypes["measures"])  # pragma: no cover
+        else:
+            dist = cdist(measures, measures)
```
Use `compute_dns` here?

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,587 @@
+            name: combined[name][survivor_indices] for name in self._store.field_list
+        }
+        self._store.clear()
+        if survivors["measures"].shape[0] > 0:
```
Repeatedly clearing and adding the store seems like it is prone to performance issues, but I can't really think of better ways at the moment. Maybe try running DNS on larger problems and we can circle back if this turns out to be a bottleneck.

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -380,47 +331,19 @@ def add(
+            for name in self._store.field_list:
+                combined[name] = (
+                    np.concatenate((cur[name], data[name]), axis=0)
+                    if name in data
```
Maybe we should throw an error instead for the case when `if name not in data`? I don't imagine your code will behave very nicely if you continue to use `combined` with some entries being longer than others.

ryanboldi replied:
Yep, addressed this too.

btjanaka commented on ribs/archives/_dns_archive.py:
I think this can be done with:

```python
survivor_indices = np.argsort(dns_scores)[-cap:]
```

Does that achieve what you intend?

ryanboldi replied:
Yes! Thanks

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -456,11 +378,23 @@ def add(
 
         # Update stats.
         if len(self) > 0:
```
Is there any case where adding will result in the archive shrinking in size? If so, this if statement would cause the archive stats to no longer be current.

ryanboldi replied:
I added an else clause that resets the stats if the Len(self) is 0 or less. Although I can't imagine why this would happen (the archive should never shrink with DNS).

btjanaka reviewed (commented):
Hi @ryanboldi, thanks for making these changes! I had a couple questions regarding your PR; could you take a look at those? **Also, please make sure to pull before making any new changes, as I made some edits to your code.** I think we are pretty close to merging!

btjanaka commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,520 @@
+        if self.empty:
+            raise IndexError("No elements in archive.")
+
+        # Deterministic selection: return the first n elites (in storage order).
```
Why does this have to be deterministic selection?

ryanboldi replied:
It doesn't! I just adjusted this to be random.

ryanboldi commented on ribs/archives/_dns_archive.py:
```diff
@@ -306,47 +258,47 @@ def index_of_single(self, measures: ArrayLike) -> Int:
 
-    def compute_dns(self, measures: ArrayLike, objectives: ArrayLike) -> np.ndarray:
-        """Computes DNS scores for a batch against the current population.
+    def _compute_dns(self, measures: ArrayLike, objectives: ArrayLike) -> np.ndarray:
```
You're right, I have adjusted this back to being a public API.

ryanboldi commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,521 @@
+        check_batch_shape(measures, "measures", self.measure_dim, "measure_dim")
+        check_finite(measures, "measures")
+
+        occupied, data = cast(
```
Adjusted this to match the ProximityArchive (no casting)

ryanboldi commented on ribs/archives/_dns_archive.py:
```diff
@@ -0,0 +1,520 @@
+    ) -> np.ndarray | BatchData | tuple[np.ndarray] | ArchiveDataFrame:
+        return self._store.data(fields, return_type)
+
+    def sample_elites(self, n: Int) -> BatchData:
```
Thanks for making this change!

ryanboldi commented:
Hey @btjanaka,

Sorry for the long delay on this! Addressed your remaining review comments:

- Field validation Now raises 'ValueError` if a field is in the archive but missing from batch data, instead of silently creating mismatched arrays
- Stats after add: Always recompute stats (including reset if archive empties), fixing the stale-stats edge case you flagged
- Deterministic → random selection: 'sample_elites' now is random (including your replace parameter)
- Removed unnecessary casts: Cleaned up where possible, kept casts where needed for ty type checking. A few of these casts needed to be added to ensure the ty checks are passed. Please advise if this is not the right approach. 

  Tests pass, ruff/ty clean. Let me know if there's anything else you'd like to add before merging!

btjanaka commented:
Hi @ryanboldi, thanks for updating! I took out the casts as I recently decided to stop using ty and use pylint instead (see #687) -- my conclusion was that ty tends to be a bit troublesome and type checking is not really worth the effort. Otherwise, the tests pass and the code looks good, so I'll be merging this once the CI finishes running.
